### PR TITLE
WIP: Autopublish IAM Roles for Service Accounts JWKS doc

### PR DESCRIPTION
**Pull request by justinsb**
This extends #9271 to automatically publish the JWKS document to a "discovery bucket", copying it from the apiserver.  We also use this for dns-controller, as a first pod, and start refactoring the IAM code to generate both node or pod roles.

Though this is "only" 2000 lines, this is an overall PR and I plan on sending smaller PRs.  But this should give the big picture!

### CODE REVIEW

**Comment by justinsb** [comment]
cc @rifelpet

File: pkg/model/components/discovery.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+	if len(kubeAPIServer.APIAudiences) == 0 {
+		kubeAPIServer.APIAudiences = []string{"kubernetes.svc.default"}
+	}
+
+	if kubeAPIServer.ServiceAccountIssuer == nil {
+		var publicURL string
+		switch p := p.(type) {
+		case *vfs.S3Path:
+			publicURL = "https://" + p.Bucket() + ".s3.amazonaws.com/" + p.Path()
```
Comment (johngmyers) [inline pkg/model/components/discovery.go]: Presumably this needs to be pushed down into the vfs interface as a new method.

File: pkg/apis/kops/v1alpha2/cluster.go [hunk 2]
```diff
@@ -191,6 +191,12 @@ type ClusterSpec struct {
 	RollingUpdate *RollingUpdate `json:"rollingUpdate,omitempty"`
 	// ServiceOIDCProvider defines the OIDC provider setup for the cluster (AWS only)
 	ServiceOIDCProvider *ServiceOIDCProviderSpec `json:"serviceOIDCProvider,omitempty"`
+
+	Discovery *DiscoverySpec `json:"discovery,omitempty"`
+}
+
+type DiscoverySpec struct {
+	Base string `json:"base,omitempty"`
```
Comment (johngmyers) [inline pkg/apis/kops/v1alpha2/cluster.go:199]: Should document this as being a VFS path.

**Comment by justinsb** [comment]
I scanned and it looks like all the functionality here is now present :-)

/close

**Pull request closed without merging**